clippy_lints/src/visibility.rs

```diff
@@ -0,0 +1,127 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, source::snippet_opt};
+use rustc_ast::ast::{Item, VisibilityKind};
+use rustc_errors::Applicability;
+use rustc_lint::{EarlyContext, EarlyLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{symbol::kw, Span};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `pub(self)` and `pub(in self)`.
+    ///
+    /// ### Why is this bad?
+    /// It's unnecessary, omitting the `pub` entirely will give the same results.
+    ///
+    /// ### Example
+    /// ```rust,ignore
+    /// pub(self) type OptBox<T> = Option<Box<T>>;
+    /// ```
+    /// Use instead:
+    /// ```rust,ignore
+    /// type OptBox<T> = Option<Box<T>>;
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub NEEDLESS_PUB_SELF,
+    complexity,
```

Comment:
I'm unsure if `complexity` is the best fit for the `needless_pub_elf` lint. While `pub(self)` does do the same as not specifying `pub` at all, it is worth considering when `pub(self)` may actually be used.

For example, as I said in the lint suggestion issue:

> Some codebases may use `pub(self)` or `pub(in self)` to denote a item that is meant for use within a submodule, and the lack of such to denote an item meant only for use within the same file.

Though I admit I didn't do any research into this point before typing that up.

Still, perhaps the lint category is worth discussing?

Reply:
There's a couple thousand usages of it it seems (though github's search is very inaccurate), most seem to just be one-off usages of it so they're either:

A: used for alignment
B: an honest mistake, as it was forgotten to be removed during refactoring (maybe if it was `pub(super` or the like before)

So, IMO, it should be `complexity`, or at least `pedantic`, as most seem to not be a stylistic choice across the codebase.

Reply:
Hmm. I'm personally unsure what to think.

Examining [the `needless_*` set of clippy lints](https://rust-lang.github.io/rust-clippy/master/index.html#/needless?groups=cargo,complexity,correctness,deprecated,nursery,pedantic,perf,restriction,style,suspicious) doesn't really help me figure out what group would be best - all those `needless_*` lints are mixed across `pedantic`, `style`, and `complexity`, and all three group sound like they may fit `needless_pub_self`.

I've tried for a bit to wrap my head around the subtleties of each group and why a particular `needless_*` lint may fit them - but it's sort of overwhelming to try to comprehend it all.

So for now, I think I'll leave this up to you and whomever else checks out this PR. My opinion is that this is a style-like lint, but I can see how it may be a `pedantic` or `complexity` lint too.

Reply:
We should ask the team what they think

Reply:
@Manishearth Thank you.

I should mention, I find it a bit weird that all the `needless_*` lints are spread out across three lint categories. Perhaps a future discussion could be dedicated to figuring out what to do with the `needless_*` lints as a whole? Whether that's reorganizing them all to one category, or providing guidance on how to categorize `needless_*` lints - because currently, they don't seem consistent in their reasoning for one category over another.

Reply:
Personally I don't really see a problem with this: the names describe the lints but there absolutely can be various kinds of "needless" things: ones where it's style consensus, ones where it's complexity, ones where it's an annoying-but-fair suggestion.

I think the default guidance applies: style for stuff that's consistent style amongst the community, complexity for stuff that is unnecessary complexity, pedantic for stuff where one of the previous applies but we think it will be too noisy and not worth it.